### PR TITLE
ceph: improve owner reference management

### PR DESCRIPTION
**Pull request by satoru-takeuchi**
**Description of your changes:**

It's better to validate ownerReferences when setting them. In addition, we should use controllerrutil.Set{Controller,Owner}Reference, that have such validation, as possible.

**Checklist:**

- [o] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [o] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [o] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [o] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [o] Documentation has been updated, if necessary.
- [o] Unit tests have been added, if necessary.
- [o] Integration tests have been added, if necessary.
- [o] Pending release notes updated with breaking and/or notable changes, if necessary.
- [o] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [o] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

**Comment by satoru-takeuchi** [comment]
@travisn This PR is what I informed you in #7185 . Could you take a look at it?

> Anyway I'll send a first PR by which reviewers can understand the idea how I will try to replace ownerReferences. I'd like to come to an agreement whether this idea make sense.

Although this PR doesn't pass unit tests and don't hanle error yet, it's enough to show how I'm going to accomplish this work.

I introduced `OwnerInfo`, that wraps both `Set{Controller,Owner}Reference` and the existing k8sutil.SetOwnerRef. There are two reasons to adopt this idea.

- It's convenient to manage owner object and `runtime.scheme` in the same struct to avoid passing these two values in many functions.
- It's impossible to get `runtime.scheme` in some functions which aren't called from controllers.

File: pkg/operator/k8sutil/resources.go [hunk 1]
```diff
@@ -24,8 +24,55 @@ import (
...
 
+// OwnerInfo is to set owner references. Only one of owner and ref must be valid.
+type OwnerInfo struct {
+	owner    metav1.Object
+	scheme   *runtime.Scheme
+	ownerRef *metav1.OwnerReference
+}
+
+// NewOwnerInfo create a new ownerInfo to set ownerReference by controllerutil
+func NewOwnerInfo(owner metav1.Object, scheme *runtime.Scheme) *OwnerInfo {
```
Comment (travisn) [inline pkg/operator/k8sutil/resources.go:40]: Tell me again why the scheme isn't always available with the owner reference? Is it when we're running inside the osd prepare job for example? Seems like everywhere else should have the scheme.

File: pkg/operator/k8sutil/resources.go [hunk 2]
```diff
@@ -24,8 +24,55 @@ import (
...
+// NewOwnerInfoWithOwnerRef create a new ownerInfo to set ownerReference by rook itself
+func NewOwnerInfoWithOwnerRef(ownerRef *metav1.OwnerReference) *OwnerInfo {
+	return &OwnerInfo{ownerRef: ownerRef}
+}
+
+// SetControllerReference set the controller reference of object
+func (info *OwnerInfo) SetControllerReference(object metav1.Object) error {
+	if info.owner != nil {
+		return controllerutil.SetControllerReference(info.owner, object, info.scheme)
+	} else {
```
Comment (travisn) [inline pkg/operator/k8sutil/resources.go:53]: nit: no need for the `else` block since the `if` block returned.

File: pkg/operator/k8sutil/resources.go (hunk 1 above)
Comment (satoru-takeuchi) [reply pkg/operator/k8sutil/resources.go:40]: There are two cases where scheme isn't available. One is osd preparae job as you said.

https://github.com/rook/rook/blob/master/cmd/rook/ceph/osd.go#L231

The other is in updating CSI driver.

https://github.com/rook/rook/blob/master/pkg/operator/ceph/operator.go#L243

Here the owner is the operator's deployment and the child is the CSI driver's configmap.

My colleague told me that we can create scheme by ourselves with `NewScheme()` and  the subsequent `AddToScheme()`s. However, it's a bit complecated and would take a bit long time. So, how about adopting the current logic for now and eliminate `k8sutil.SetOwnerRef()` after that?

Comment (travisn) [reply pkg/operator/k8sutil/resources.go:40]: In two steps is fine. The key now is to have thorough testing to prevent regressions. With such a big change to the owner references, let's also test by ensuring the K8s garbage collector triggers rebuilding its index. This will help us verify resources don't later disappear 24 hours after everything is fine. However, I don't recall how to force trigger the GC rebuild.

Comment (satoru-takeuchi) [reply pkg/operator/k8sutil/resources.go:40]: > let's also test by ensuring the K8s garbage collector triggers rebuilding its index.

Let me clarify what you mean. Isn't it enough to already merged `kubectl-check-ownerreferences`. If there are invalid ownerReferences, this program detect it. Na resource deletion only happenss due to invalid references.

>  I don't recall how to force trigger the GC rebuild.

Me neither. I guess how to force trigger the GC rebuild depends on implementation, so K8s version. So it's dufficult to do it.

Comment (travisn) [reply pkg/operator/k8sutil/resources.go:40]: > > let's also test by ensuring the K8s garbage collector triggers rebuilding its index.
> 
> Let me clarify what you mean. Isn't it enough to already merged `kubectl-check-ownerreferences`. If there are invalid ownerReferences, this program detect it. Na resource deletion only happenss due to invalid references.

Oh right, that CI action will be perfect to detect the issues. I haven't seen that action fail yet in any PRs. Have you seen it fail? If we could just double check that the CI action fails when there is a real failure, I think we're covered, thanks.

Comment (satoru-takeuchi) [reply pkg/operator/k8sutil/resources.go:40]: > Have you seen it fail?

No. After making this PR ready to be reviewed, I'll check it again.

File: pkg/operator/k8sutil/resources.go (hunk 2 above)
Comment (satoru-takeuchi) [reply pkg/operator/k8sutil/resources.go]: fixed, thanks.

**Comment by satoru-takeuchi** [comment]
The remaining work is to handle errors properly and to pass integration tests.

**Comment by satoru-takeuchi** [comment]
This PR is ready to be reviewed. All tests passed.

Note:
I'd like to merge this PR and one more subsequent PR in the v1.6 development.
It's because that major release is a good time to merge such big changes: this
PR is over 3,000 loc and the other PR would also consist of more than 1,000 loc.
By merging these two PRs, all settings of ownerReference are verified on-time.

File: pkg/operator/ceph/cluster/osd/spec_test.go [hunk 3]
```diff
@@ -29,15 +29,22 @@ import (
...
 	"github.com/stretchr/testify/assert"
 	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/client-go/kubernetes/fake"
 )
 
 func TestPodContainer(t *testing.T) {
 	cluster := &Cluster{rookVersion: "23", clusterInfo: client.AdminClusterInfo("myosd")}
+	cephCluster := cephv1.CephCluster{}
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/spec_test.go:43]: It seems adding this reference in the tests is a lot of duplicate code. Could we have a helper method to set it in the test? Or is there a pattern used in the other tests that could be followed here? I don't see the duplication in the other unit tests like this one.

**Review by travisn — APPROVED** [review]
Agreed, we need to get this stabilizing in master. I ran a quick test and looks good to me. Just one small suggestion for the tests, but I'll approve so you can go merge when ready.

File: pkg/operator/ceph/cluster/osd/spec_test.go (hunk 3 above)
Comment (satoru-takeuchi) [reply pkg/operator/ceph/cluster/osd/spec_test.go:43]: It's introduced in this PR. The original code just needs empty ownerReference but now one needs ownerInfo with setting some fields in many cases. I'll add some helper scripts.
